dace/libraries/blas/nodes/dot.py

```diff
@@ -62,51 +68,43 @@ def make_sdfg(node, parent_state, parent_sdfg):
         init_state.add_memlet_path(init_tasklet,
                                    init_write,
                                    src_conn="_out",
-                                   memlet=dace.Memlet.simple(init_write.data,
-                                                             "0",
-                                                             num_accesses=1))
+                                   memlet=dace.Memlet("_result[0]"))
 
         # Multiplication map
         state.add_mapped_tasklet(
             "_DOT_", {"__i": "0:{}".format(N)}, {
-                "__x": dace.Memlet.simple("_x", "__i"),
-                "__y": dace.Memlet.simple("_y", "__i")
-            },
-            mul_program, {
-                "__out":
-                dace.Memlet.simple(mul_out, "0", wcr_str="lambda x, y: x + y")
+                "__x": dace.Memlet("_x[__i]"),
+                "__y": dace.Memlet("_y[__i]")
             },
+            mul_program,
+            {"__out": dace.Memlet(f"{mul_out}[0]", wcr="lambda x, y: x + y")},
             external_edges=True,
             output_nodes=output_nodes)
 
         return sdfg
 
-    @staticmethod
-    def expansion(node, state, sdfg):
-        node.validate(sdfg, state)
-        if node.dtype is None:
-            raise ValueError("Data type must be set to expand " + str(node) +
-                             ".")
-        return ExpandDotPure.make_sdfg(node, state, sdfg)
-
 
 @dace.library.expansion
 class ExpandDotOpenBLAS(ExpandTransformation):
 
     environments = [environments.openblas.OpenBLAS]
 
     @staticmethod
-    def expansion(node, state, sdfg):
-        node.validate(sdfg, state)
-        dtype = node.dtype
+    def expansion(node, state, sdfg, n=None, **kwargs):
+        desc_x, _, _ = node.validate(sdfg, state)
+        dtype = desc_x.dtype.base_type
+        veclen = desc_x.dtype.veclen
         if dtype == dace.float32:
             func = "sdot"
         elif dtype == dace.float64:
             func = "ddot"
         else:
             raise ValueError("Unsupported type for BLAS dot product: " +
                              str(dtype))
-        code = "_result = cblas_{}(n, _x, 1, _y, 1);".format(func)
+        n = n or node.n
+        if veclen != 1:
+            n /= veclen
+        code = f"_result = cblas_{func}({n}, _x, 1, _y, 1);"
```

Comment:
This is something that is done inconsistently across our nodes, and I didn't want to change it yet until we've talked about the best way of actually doing this (with Alexnick also). Using the incoming memlets is not always possible because of streams, where the memlet subsets are all 1D. Furthermore, it's not necessarily obvious to map subsets to dimensions once you read the same container multiple times. Are you okay with leaving this for now, then we have a discussion on how to do this best with Alexnick's use cases?

Reply:
yep